### PR TITLE
Remove support for `associated_type_bound` nested in `dyn` types

### PR DESCRIPTION
**Pull request by compiler-errors**
These necessarily desugar to `impl Trait`, which is inconsistent with the `associated_type_bound` feature after #120584.

This PR keeps the `is_in_dyn_type` hack, which kind of makes me sad. Ideally, we'd be validating that no object types have associated type bounds somewhere else. Unfortunately, we can't do this later during astconv (i think?), nor can we do it earlier during ast validation (i think?) because of the feature gating of ATB being a *warning* rather than an *error*. Let me know if you have thoughts about this.

r? lcnr

### CODE REVIEW

File: compiler/rustc_ast_lowering/messages.ftl [hunk 1]
```diff
@@ -8,6 +8,10 @@ ast_lowering_arbitrary_expression_in_pattern =
 
 ast_lowering_argument = argument
 
+ast_lowering_assoc_ty_binding_in_dyn =
+    associated type bounds are not allowed in `dyn` types
+    .suggestion = use `impl` to introduce a type instead
```
Comment (lcnr) [inline compiler/rustc_ast_lowering/messages.ftl:13]:
```suggestion
    .suggestion = use `impl Trait` to introduce a type instead
```

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -1085,14 +1085,32 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                            .as_ref()
+                            .map_or(constraint.ident.span, |args| args.span());
+                        let colon_span = if bound_end_span.eq_ctxt(constraint.span) {
+                            Some(self.tcx.sess.source_map().next_point(bound_end_span))
+                        } else {
+                            None
+                        };
+                        DesugarKind::Error(colon_span)
+                    }
+                    _ if self.is_in_dyn_type => DesugarKind::Error(None),
```
Comment (lcnr) [inline compiler/rustc_ast_lowering/src/lib.rs:1112]: nit: can you change this to first check `self.is_in_dyn_type` and then match on `itctx` inside of that? :sweat_smile: was a bit confusing for me as is

**Review by lcnr — COMMENTED** [review]
I think `is_in_dyn_type` is just a more specific version of "in type". So maybe just changing this to "in_type" seems a bit nicer as it's *more general*, but really :shrug:

**Comment by lcnr** [comment]
r=me after nits

**Comment by compiler-errors** [comment]
I'm gonna land this after #120584, since it conflicts.

**Comment by compiler-errors** [comment]
Removing `DesugarKind::ImplTrait` caused us to have an unused fn, removed that.

@bors r=lcnr

**Pull request merged**